command/agent/exec/exec.go

```diff
@@ -0,0 +1,288 @@
+	for id, ctmpls := range idMap {
+		for _, ctmpl := range ctmpls {
+			tl := lookupMap[id]
+			tl = append(tl, ctmpl)
+			lookupMap[id] = tl
+		}
+	}
+	s.lookupMap = lookupMap
+
+	for {
+		select {
+		case <-ctx.Done():
+			s.runner.Stop()
+			s.procLock.Lock()
+			if s.proc != nil {
+				s.proc.Stop()
+			}
+			s.procStarted = false
+			s.proc.Unlock()
+			return nil
+		case token := <-incomingVaultToken:
+			if token != *latestToken {
+				s.logger.Info("exec server received new token")
+
+				s.runner.Stop()
+				*latestToken = token
+				ctv := ctconfig.Config{
+					Vault: &ctconfig.VaultConfig{
+						Token:           latestToken,
+						ClientUserAgent: pointerutil.StringPtr(useragent.AgentTemplatingString()),
+					},
+				}
+
+				runnerConfig = runnerConfig.Merge(&ctv)
+				var runnerErr error
+				s.runner, runnerErr = manager.NewRunner(runnerConfig, true)
+				if runnerErr != nil {
+					s.logger.Error("template server failed with new Vault token", "error", runnerErr)
+					continue
+				}
+				go s.runner.Start()
+			}
+
+		case err := <-s.runner.ErrCh:
+			s.logger.Error("template server error", "error", err.Error())
+			s.runner.StopImmediately()
+
+			// Return after stopping the runner if exit on retry failure was specified
+			if s.config.AgentConfig.TemplateConfig != nil && s.config.AgentConfig.TemplateConfig.ExitOnRetryFailure {
+				return fmt.Errorf("template server: %w", err)
+			}
+
+			s.runner, err = manager.NewRunner(runnerConfig, true)
+			if err != nil {
+				return fmt.Errorf("template server failed to create: %w", err)
+			}
+			go s.runner.Start()
+		case <-s.runner.TemplateRenderedCh():
+			// A template has been rendered, figure out what to do
+			s.logger.Debug("template rendered")
+			events := s.runner.RenderEvents()
+
+			// events are keyed by template ID, and can be matched up to the id's from
+			// the lookupMap
+			if len(events) < len(s.lookupMap) {
+				// Not all templates have been rendered yet
+				continue
+			}
+
+			// assume the renders are finished, until we find otherwise
+			doneRendering := true
+			envVarToContents := map[string]string{}
+			for _, event := range events {
+				// This template hasn't been rendered
+				if event.LastWouldRender.IsZero() {
+					doneRendering = false
+				} else {
+					// TODO: check for duplicates?
+					for _, tcfg := range event.TemplateConfigs {
+						envVarToContents[*tcfg.MapToEnvironmentVariable] = string(event.Contents)
+					}
+				}
+			}
+
+			if doneRendering {
+				s.logger.Debug("done rendering templates/detected change, bouncing process")
+				if err := s.bounceCmd(envVarToContents); err != nil {
+					return fmt.Errorf("unable to bounce command: %w", err)
+				}
+			}
+		case exitCode := <-s.exitCh:
+			// process exited on its own
+			return &ProcessExitError{ExitCode: exitCode}
+		}
+	}
+}
+
+func (s *Server) bounceCmd(newEnvVars map[string]string) error {
+	s.procLock.Lock()
+	defer s.procLock.Unlock()
```

Comment:
Which go-routines are we protecting against? It seems like everything is happening within one goroutine (Run) so it's probably OK to get rid of the lock.

Reply:
If thats the case how does not seize up currently on a lock or is the go-routine for Run not locking/unlocking?

Reply:
> there can be an update to a secret while child process is in the middle of restarting (see s.runner.TemplateRenderedCh)

When an update is delivered on this channel, I believe it will block waiting for whatever is being processed currently by `Run`. So if Run is currently restarting the process, any update(s) will be blocked till that is done with or without this lock since Run is single-threaded.

Reply:
> If thats the case how does not seize up currently on a lock or is the go-routine for Run not locking/unlocking?

Locking/unlocking should be a no-op in single-threaded context.

Reply:
> When an update is delivered on this channel, I believe it will block waiting for whatever is being processed currently by Run. So if Run is currently restarting the process, any update(s) will be blocked till that is done with or without this lock since Run is single-threaded.

OK, this makes sense. i'll remove the lock